packages/astro/src/core/render/dev/index.ts

```diff
@@ -80,7 +81,42 @@ export async function preload({
 	const renderers = await loadRenderers(viteServer, astroConfig);
 	// Load the module from the Vite SSR Runtime.
 	const mod = (await viteServer.ssrLoadModule(fileURLToPath(filePath))) as ComponentInstance;
-
+	if (viteServer.config.mode === 'development' || !mod?.$$metadata) {
+		return [renderers, mod];
+	}
+	// During prod builds, some modules have dependencies we should preload by hand
+	// Ex. markdown files imported asynchronously or via Astro.glob(...)
+	// We will call each md file's $$loadMetadata to discover those dependencies
+	const modGraph = await viteServer.moduleGraph.getModuleByUrl(fileURLToPath(filePath));
+	for await (let importedModule of modGraph?.importedModules ?? []) {
+		if (importedModule.id?.endsWith(MARKDOWN_IMPORT_FLAG)) {
+			const mdMod = await viteServer.ssrLoadModule(importedModule.id);
+			const mdMetadata = (await mdMod.$$loadMetadata?.()) as Metadata;
+			if (mdMetadata) {
+				// if md metadata, merge with parent page metadata
+				const { modules, hoisted, hydratedComponents, clientOnlyComponents, hydrationDirectives } =
+					mod.$$metadata;
+				mod.$$metadata.modules = modules.concat(
```

Comment:
Instead of doing this immutable stuff it would be more efficient to just push the changes into the arrays. You are already modifying them.

Reply:
Fixed!